model/validation.go

```diff
@@ -0,0 +1,907 @@
+package model
+
+// This file contains methods for validating a role manifest
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strings"
+
+	"code.cloudfoundry.org/fissile/validation"
+)
+
+// validateVariableDescriptions tests whether all variables have descriptions
+func validateVariableDescriptions(roleManifest *RoleManifest) validation.ErrorList {
+	allErrs := validation.ErrorList{}
+
+	for _, variable := range roleManifest.Variables {
+		if variable.CVOptions.Description == "" {
+			allErrs = append(allErrs, validation.Required(variable.Name,
+				"Description is required"))
+		}
+	}
+
+	return allErrs
+}
+
+// validateSortedTemplates tests that all templates are sorted in alphabetical order
+func validateSortedTemplates(roleManifest *RoleManifest) validation.ErrorList {
+	allErrs := validation.ErrorList{}
+
+	previousKey := ""
+
+	for _, templateDef := range roleManifest.Configuration.Templates {
+		key := templateDef.Key.(string)
+
+		if previousKey != "" && previousKey > key {
+			allErrs = append(allErrs, validation.Forbidden(previousKey,
+				fmt.Sprintf("Template key does not sort before '%s'", key)))
+		}
+
+		previousKey = key
+	}
+
+	return allErrs
+}
+
+// validateScripts tests that all referenced scripts exist, and that all scripts
+// are referenced.
+func validateScripts(roleManifest *RoleManifest) validation.ErrorList {
+	allErrs := validation.ErrorList{}
+	roleManifestDirName := filepath.Dir(roleManifest.manifestFilePath)
+	scriptsDirName := filepath.Join(roleManifestDirName, "scripts")
+	usedScripts := map[string]bool{}
+	err := filepath.Walk(scriptsDirName, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if strings.HasPrefix(info.Name(), ".") {
+			if info.IsDir() {
+				return filepath.SkipDir // No need to walk hidden directories
+			}
+			return nil // Ignore all hidden files
+		}
+		if info.IsDir() {
+			return nil // Ignore directories, but recurse into them
+		}
+
+		relpath, err := filepath.Rel(scriptsDirName, path)
+		if err != nil {
+			return err
+		}
+		usedScripts["scripts/"+relpath] = false
+		return nil
+	})
+	if err != nil && !roleManifest.validationOptions.AllowMissingScripts {
+		return append(allErrs, validation.Invalid(scriptsDirName, err.Error(), "Error listing files in scripts directory"))
+	}
+
+	for _, instanceGroup := range roleManifest.InstanceGroups {
+		for scriptType, scriptList := range map[string][]string{
+			"script":             instanceGroup.Scripts,
+			"environment script": instanceGroup.EnvironScripts,
+			"post config script": instanceGroup.PostConfigScripts,
+		} {
+			for _, script := range scriptList {
+				if filepath.IsAbs(script) {
+					continue
```

Comment:
Do we actually allow absolute script paths ?
Should we maybe print a warning about the trouble this can be, even if not an error ?

IMHO this is special enough to require an informational comment.

Reply:
We do! Because (at least at some point) we manually added scripts from `/var/vcap/packages`, which fissile has no knowledge about.

Reply:
Ok. Can we document that, please?

Reply:
Comment added in 97c7c9d; let me know if you have a different kind of document in mind.

Reply:
Your comment looks ok.